src/thor/route_action.cc

```diff
@@ -590,7 +642,52 @@ void thor_worker_t::path_depart_at(Api& api, const std::string& costing) {
     // if we just made a leg that means we are done recording which algorithms were used
     if (path.empty())
       algorithms.clear();
+
+    return true;
+  };
+
+  auto correlated = options.locations();
+  bool allow_retry = true;
+
+  // For each pair of locations
+  auto destination = ++correlated.begin();
+  while (destination != correlated.end()) {
+    auto origin = std::prev(destination);
+    if (!route_two_locations(origin, destination)) {
+      // if routing failed because an intermediate waypoint was snapped to the low reachability road
+      // (such road lies in a small connectivity component that is not connected to other locations)
+      // we should leave only high reachability candidates and try to route again
+      if (allow_retry && origin != correlated.begin() && is_through_point(*origin) &&
+          origin->path_edges_size() > 0 && !is_highly_reachable(*origin, origin->path_edges(0))) {
```

Comment:
ok so we dont have to solve this in this PR but... it is possible that the router will use a different edge at the end of one leg than the beginning of the next one. so while you were right that we only need to currently care about handling a failure for through locations (because only they widdle down the edges to just the one that was used in the adjacent leg) it is the case that the router could have a situation like i described in my previous comment.

lets go with the simple case: break->break->break

assume that the middle break point has two pairs of edge candidates one inside a small component and one outside. assume also that the first break point is inside the small component. what happens is then the fist leg is in the component and the second leg is outside of it. makign the returned route, invalid. so you are right, it wont fail the route but it will return something unusable to the user. this is a long standing issue and a rare occurance, but i wanted to point it out to you here.

we probably should be removing some of the pairs of candidates even for breaks and vias if they arent directly connected to the last edge in the previous edge. thats i think what i planned to do with that commented out function above but it was so long ago i dont quite remember ))

Reply:
it's very useful comment, thank you! we should rethink this logic and fix in a separate issue. I'll crate after this one gets merged